core/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -215,22 +229,61 @@ void dispatchRequest(final RestRequest request, final RestChannel channel, final
+                deprecationLogger.deprecated("Plain text request bodies are deprecated. Use request parameters or body " +
+                    "in a supported format.");
+            } else if (isContentTypeRequired) {
+                return false;
+            } else {
+                deprecationLogger.deprecated("Content type detection for rest requests is deprecated. Specify the content type using" +
+                    "the [Content-Type] header.");
+                XContentType xContentType = XContentFactory.xContentType(restRequest.content());
+                if (xContentType == null) {
+                    if (restHandler != null) {
+                        if (restHandler.supportsPlainText()) {
+                            deprecationLogger.deprecated("Plain text request bodies are deprecated. Use request parameters or body " +
+                                "in a supported format.");
+                        } else {
+                            return false;
+                        }
+                    }
+                } else {
+                    restRequest.setXContentType(xContentType);
+                }
+            }
+        }
+        return true;
+    }
+
+    private void sendContentTypeErrorMessage(RestRequest restRequest, RestChannel channel) throws IOException {
+        final List<String> contentTypeHeader = restRequest.getAllHeaderValues("Content-Type");
+        final String errorMessage;
+        if (contentTypeHeader == null) {
+            errorMessage = "Content-Type header is missing";
+        } else {
+            errorMessage = "Content-Type header [" +
+                Strings.collectionToCommaDelimitedString(restRequest.getAllHeaderValues("Content-Type")) + "] is not supported";
+        }
+
+        channel.sendResponse(BytesRestResponse.createSimpleErrorResponse(NOT_ACCEPTABLE, errorMessage));
+    }
+
+    /**
      * Checks the request parameters against enabled settings for error trace support
      * @return true if the request does not have any parameters that conflict with system settings
      */
     boolean checkRequestParameters(final RestRequest request, final RestChannel channel) {
         // error_trace cannot be used when we disable detailed errors
         // we consume the error_trace parameter first to ensure that it is always consumed
         if (request.paramAsBoolean("error_trace", false) && channel.detailedErrorsEnabled() == false) {
-            try {
-                XContentBuilder builder = channel.newErrorBuilder();
-                builder.startObject().field("error", "error traces in responses are disabled.").endObject().string();
-                RestResponse response = new BytesRestResponse(BAD_REQUEST, builder);
-                response.addHeader("Content-Type", "application/json");
-                channel.sendResponse(response);
-            } catch (IOException e) {
-                logger.warn("Failed to send response", e);
-            }
```

Comment:
right....with all this folding business in github ctrl+F doesn't always work :(  it does simplify things thanks